### ISSUE TITLE
PHP WARNING: "array_merge(): Argument #2 is not an array" in file engine/classes/Elgg/Database/Annotations.php (line 372)

### ISSUE THREAD

**Issue by rohit1290** [Elgg/Elgg#9785]
I noted that the above mentioned PHP warning is appearing in my error log file. I inspected the array_merge() function at line 372 of Annotations.php file and saw that the 2nd argument is $time_wheres and _elgg_get_entity_time_where_sql is returning a "string" to $time_wheres parameter at line 368 of Annotations.php file.

This can be solved by changing line 368 of Annotations.php file from

> $time_wheres = _elgg_get_entity_time_where_sql('a', $options['annotation_created_time_upper'], $options['annotation_created_time_lower']);

	
to

> $time_wheres = **array(**_elgg_get_entity_time_where_sql('a', $options['annotation_created_time_upper'], $options['annotation_created_time_lower'])**)**;

I am using Elgg 1.12.x

**Comment by mrclay** [comment]
On what page does this error occur? I want to make sure this is not due to incorrect API usage.

**Comment by rohit1290** [comment]
I am not sure which page is causing this error. I just got the log in my error log file. 
How can I know which page is causing this error?

**Comment by mrclay** [comment]
I propose a different fix: #9787

**Comment by mrclay** [comment]
We should not accept any fix without a basic test to exercise it. As no one can locate what it affects, I don't see urgency to rush a fix.

**Comment by rohit1290** [comment]
The fix that you proposed is doing the work of array_merge and that too with a string. I have done a basic testing to see the output and found the fix valid. https://github.com/Elgg/Elgg/pull/9787#discussion_r62029169

I have even made the correction on my website 2 days ago and have not encountered the above mentioned error till now.

**Comment by rohit1290** [comment]
@mrclay, I tested your fix #9787 and found that it started generating the below error:

https://elgg.org/discussion/view/2494740/exception-databaseexception-with-message-unknown-column-atime-created-in-where-clause

**Comment by iionly** [comment]
@mrclay , @rohit1290 proposed alternative fix:
```php
		$time_wheres = _elgg_get_entity_time_where_sql('n_table', $options['annotation_created_time_upper'],
			$options['annotation_created_time_lower']);
	
		if ($time_wheres) {
			$time_wheres = explode(' AND ', $time_wheres);
			$options['wheres'] = array_merge($options['wheres'], $time_wheres);
		}
```
I don't know how to write tests. I also don't know why the 'a' table is now expected as 'n_table'. I think it only doesn't crash more often at the moment because the $time_wheres might not contain anything in most cases. The function  _elgg_get_entity_time_where_sql() calls the function getEntityTimeWhereSql(). In getEntityTimeWhereSql() the where clauses are imploded with ' AND ' and a string is returned. Therefore I think it should be okay to simply explode the string again before merging with the $options['wheres'] array.

**Comment by mrclay** [comment]
New fix with test #9807

**Comment by rohit1290** [comment]
Fix https://github.com/Elgg/Elgg/pull/9807 is not creating any further issue. 
Both the array_merge warning and databaseexception error is solved from this fix.
